pulser/simulation/simulation.py

```diff
@@ -147,34 +245,77 @@ def prepare_dict() -> dict[str, np.ndarray]:
                     'phase': np.zeros(self._tot_duration)}
 
         def write_samples(slot: _TimeSlot,
-                          samples_dict: Mapping[str, np.ndarray]) -> None:
+                          samples_dict: Mapping[str, np.ndarray],
+                          *qid: Union[int, str],
+                          is_global_pulse: bool = False) -> None:
+            """Builds hamiltonian coefficients.
+
+            Taking into account, if necessary, noise effects, which are local
+            and depend on the qubit's id qid.
+            """
             _pulse = cast(Pulse, slot.type)
-            samples_dict['amp'][slot.ti:slot.tf] += _pulse.amplitude.samples
-            samples_dict['det'][slot.ti:slot.tf] += _pulse.detuning.samples
+            noise_det = 0.
+            noise_amp = 1.
+
+            for noise in self.config.noise:
+                if noise == 'doppler':
+                    noise_det += np.random.normal(
+                        0, self.config.doppler_sigma)
+                # Gaussian beam loss in amplitude for global pulses only
+                if noise == 'amplitude' and is_global_pulse:
+                    position = self._qdict[qid[0]]
+                    r = np.linalg.norm(position)
+                    w0 = self.config.laser_waist
+                    noise_amp = np.random.normal(
+                        1., 1.e-3) * np.exp(-(r/w0)**2)
+
+            samples_dict['amp'][slot.ti:slot.tf] = \
+                _pulse.amplitude.samples * noise_amp
+            samples_dict['det'][slot.ti:slot.tf] = \
+                _pulse.detuning.samples + noise_det
```

Comment:
So, it seems that the only difference is that you multiply the samples by the noise parameters here, which are randomly sampled? What if you stored the clean samples_dict once and then, on each run, qubit and pulse, recalculated the `noise_amp` and `noise_det` to get the noisy samples dict? That might save a bit of time, no?

Reply:
Alright, that might save time. You mean having two different `clean_sample_dict`and `noisy_sample_dict`?

Reply:
You store the `clean_sample_dict` once, and then can create the "current" sample dict from it. The `clean_sample_dict` remains unchanged, and you keep updating `noisy_sample_dict` whenever necessary.

Reply:
Okay, that's what I had in mind.

Reply:
Trying to implement this, this actually doesn't change much. We still need to go through each channel and slot to add noises at correct slot times (if there's no laser, detuning shouldn't appear in the Hamiltonian in the doppler case). It would have been a nice boost if we could just change all the samples acting on a qubit all at once but this isn't the case... I think I'll stick with the old method. That said, this made me realise a design flaw of the old code when dealing with multiple pulses : I randomly chose the velocity of the atom at each pulse, which isn't realistic since we consider that velocity is constant throughout a sequence. I'll make the changes.

Reply:
It is true that if you still need to do the full loop, it probably won't help much.